[internal] terraform: support consuming terraform from local path

tdyas opened the pull request:
Allow using the `terraform` binary found at the path specified by `--terraform-path` instead of downloading Terraform from the Internet.

[ci skip-rust]

[ci skip-build-wheels]

tdyas commented:
My main questions arei:
- What is/are the right option(s) for overriding `ExternalTool` binaries with a local path?
- Should there be an option to discover the binary on the PATH?

stuhood commented:
Although we need to do this for go, do you think that it is actually worthwhile for terraform?
* Does it make a noticeable difference in performance?
* Do we think that users are likely to need custom built terraform binaries?

If not, then it seems to add unnecessary complexity.

Eric-Arellano commented:
Agreed. This makes remote caching less useful and increases the risk of Works On My Machine type errors. 

I don't like that we have to use PATH for Go: it's meant as a pragmatic compromise because immutable caches will take too much time. Once we have immutable caches, I was hoping to remove our PATH support.

tdyas commented:
> Do we think that users are likely to need custom built terraform binaries?

They may not need a custom-built binary, but rather a standard Terraform binary installed into a standard environment (e.g., Docker image). Thus, searching the `PATH` or providing an explicit path allows consumption of a Terraform binary from that standard environment.

@asherf: Thoughts on this issue?

asherf commented:
> > Do we think that users are likely to need custom built terraform binaries?
> 
> They may not need a custom-built binary, but rather a standard Terraform binary installed into a standard environment (e.g., Docker image). Thus, searching the `PATH` or providing an explicit path allows consumption of a Terraform binary from that standard environment.
> 
> @asherf: Thoughts on this issue?

I am for it, but I would make it an `advanced` option and maybe document the risk.
The main reason I think this is advantageous is the current mechanism pants has for adding tool versions, which requires adding each version multiple times (for each platform) with the size & checksum - I did that last week, currently it is manual & cumbersome process. Considering how often TF is released, using TF from the path would make it easier to upgrade.

tdyas commented:
Updated:
- Renamed `--download-terraform` subsystem to just `--terraform`.
- Added `--terraform-search-paths` option.
- Logic to find the binary now uses `--terraform-path` if set, then tries searching `--terraform-search-paths` if set, otherwise falls back to downloading Terraform.

I'd like input on the options and how we should differentiate between the cases (explicit path, searching paths, and downloading).

Eric-Arellano commented:
Given the insight at https://github.com/pantsbuild/pants/pull/13023#issuecomment-929233655, I wonder if we should ever be downloading Terraform for you? That is, we know you cannot use Terraform entirely through Pants - we will likely only do linting/formatting for you. You still need it installed in the system for anything else.

It's important that the version of system installed is == what Pants uses. @asherf has pointed out how tedious it is to do that with Pants downloading because of the checksum validation, and that's also installing the same binary twice.

Instead, I think only using system install makes sense. Like we do in https://github.com/pantsbuild/pants/pull/13007, the user sets the expected TF version and we iterate over all binaries until finding a valid one. 

cc @kaos @benjyw @stuhood

tdyas commented:
Should it use a semver constraint instead of an expected version?

Eric-Arellano commented:
> Should it use a semver constraint instead of an expected version?

That could be cool! But I'm not sure it's necessary given project priorities? Fwit, Go is ignoring patch versions and only doing major.minor:

https://github.com/pantsbuild/pants/blob/537b2e72c0aa971a223b8a19d08ef0bccafc92cf/src/python/pants/backend/go/subsystems/golang.py#L124-L128

benjyw commented:
> Should it use a semver constraint instead of an expected version?

Generally isn't Terraform picky about data compatibility, and thus insistent on exact versions?

tdyas commented:
> > Should it use a semver constraint instead of an expected version?
> 
> Generally isn't Terraform picky about data compatibility, and thus insistent on exact versions?

It can be which is why having an exact match to a specific version down to patch-level is probably the better path. A version constraint would provide that versus just matching on the major/minor versions.

tdyas commented:
And just matching down to patch-level exactly would also provide that. At Toolchain, we pin Terraform to exact versions.

tdyas commented:
Switched the PR to use search paths and a semver specifier set (e.g., `>=1.0,<2`).

Tests pass locally but will fail in CI since there is not a `terraform` binary installed. How should we handle that?

Separate CI shard that has `terraform` installed?

Eric-Arellano commented on src/python/pants/backend/terraform/tool.py:
```diff
@@ -2,44 +2,165 @@
+
+        invalid_versions.append((binary_path.path, version))
+
+    bulleted_list_sep = "\n  * "
+    invalid_versions_str = bulleted_list_sep.join(
+        f"{path}: {version}" for path, version in sorted(invalid_versions)
+    )
```
This should use `bullet_list` from `pants.util.strutil`, landed in the past day. Check `golang.py` for an example

Eric-Arellano commented on src/python/pants/backend/terraform/tool.py:
```diff
@@ -53,21 +174,12 @@ class TerraformProcess:
     return Process(
-        argv=("./terraform",) + request.args,
-        input_digest=input_digest,
+        argv=(terraform_setup.path.path,) + request.args,
```
```suggestion
        argv=(terraform_setup.path.path, *request.args),
```

Eric-Arellano reviewed (approved):
Thanks! CI should install Terraform imo in the Linux test shard, but not the other ones.

stuhood commented on src/python/pants/backend/terraform/tool.py:
```diff
@@ -2,44 +2,165 @@
+    options_scope = "terraform"
     help = "Terraform (https://terraform.io)"
 
-    default_version = "1.0.7"
-    default_url_template = (
-        "https://releases.hashicorp.com/terraform/{version}/terraform_{version}_{platform}.zip"
```
So, even if we support using this from the PATH, it seems like we should still keep the support for fetching it? It certainly makes it easier for us as consumers to be able to run hermetically from an empty path.

Eric-Arellano replied:
I argue in https://github.com/pantsbuild/pants/pull/12955#issuecomment-929283138 why I think we should "keep it simple"

Eric-Arellano replied:
cc @benjyw and @kaos, I don't think we've heard your opinion yet on whether we should support for Terraform 1) only use from local PATH, 2) only use from Pants downloading, or 3) both.

This decision also will apply to Go, which for now only supports local PATH. When we add "immutable caches" so that Pants downloading is performant, we'll need to decide if we still keep local PATH support or not.

stuhood replied:
I think that it makes contributors lives easier to bootstrap this for you, and that in consumer respositories there will be users who might "use the repository" without actually executing terraform directly for sideeffects. Basically: the convenience provided here for our contributors can be mirrored in user repositories as well.

So my vote would in general be to support both bootstrapping and using a system install (with a `--system` toggle) where possible. It seems like we can't for `go` (yet!) for practical reasons, but this might apply there as well in future.

Eric-Arellano replied:
Okay, that's sensible enough to me. 

We may need to tweak the options interface for `ExternalTool` to work better with installing from PATH. For example, `--version` might not as useful, given Tom's recommendation that we should support semantic versions. All the users of this pattern (Java, Go, Terraform) are still experimental, so we can punt on that redesign - but it'd be good to open a ticket and figure out before we stabilize these backends.

tdyas replied:
So is the consensus to use ExternalTool's option plus these options (using `terraform` as an example): `--terraform-system`, `--terraform-search-paths`, and `--terraform-version-constraint` ?

Eric-Arellano replied:
In my comment above, I speculate that we probably want to rename the `ExternalTool` options if system installs are supported. Rather than `--terraform-version` (downloads) and `--terraform-version-constraint` (system), we probably want something like `--terraform-download-version` and `--terraform-system-version-constraint`.

But I think we can punt on redesigning the option names, so long as we have a ticket tracking it.

tdyas commented:
Opened a thread on Pants Slack about whether target types might make sense for tool configuration: https://pantsbuild.slack.com/archives/C0D7TNJHL/p1633468006077300

Going to see how that discussion goes before proceeding with this PR.
